draft-ietf-masque-connect-udp-listen.md

```diff
@@ -154,15 +176,118 @@ UDP Payload:
-for the Connect-UDP-Bind header field value, but future documents might define
-parameters. Receivers MUST ignore unknown parameters.
+
+### Compression Mapping {#mappings}
+
+When an endpoint receives a COMPRESSION_ASSIGN capsule with a non-zero IP length, it MUST decide whether to accept or reject the compression mapping:
+
+if it accepts the mapping, first the receiver MUST save the mapping from context ID to address and port. Second, the receiver MUST echo an identical COMPRESSION_ASSIGN capsule back to its peer.
+
+if it rejects the mapping, the receiver MUST respond by sending a COMPRESSION_CLOSE capsule with the context ID set to the one from the received COMPRESSION_ASSIGN capsule
+
+The client or proxy MAY choose to close any context contexts that it registered
+or was registered with it respectively using COMPRESSION_CLOSE
+(For example when a mapping is unused for a long time). Another potential use is
+{{restrictingips}}.
+
+
+## Restricting IPs {#restrictingips}
+
+If an uncompressed Context ID was set (via {{uncompressed}}), the client MAY at any point request the proxy reject all traffic from uncompressed
+targets by using COMPRESSION_CLOSE (see {{compressionclose}}) on said Context ID.
+targets effectively acting as a firewall against unwanted or unknown IPs.
+
+
+## Capsules {#capsules}
+The Listener capsule types are defined as follows:
+
+### The COMPRESSION_ASSIGN capsule {#compressionassign}
+
+The Compression Assign capsule has two purposes. Either to request the assignment of a Context ID (see {{contextid}}) to a corresponding target IP:Port. Or to accept a COMPRESSION_ASSIGN request from the other party.
+
+~~~ ascii-art
+Capsule {
+  Type COMPRESSION_ASSIGN (0x1F1F1F1F),
+  Length (i),
+  Target Information,
+}
+~~~
+{: #capsuleassignformat title="Compression Assign Capsule Format"}
+
+~~~ ascii-art
+Target Information {
+  Context ID (i),
+  IP Version (8),
+  IP Address (32..128),
+  UDP Port (16),
+}
+~~~
+{: #targetmappingformat title="Target Information Format"}
+
+The IP Length, Address and Port fields in {{targetmappingformat}} are the
+same as those defined in {{format}}. However, the IP version can be set
+to 0 when allocating an uncompressed Context ID, as defined in {{contextid}}.
+
+### The COMPRESSION_CLOSE capsule {#compressionclose}
+
+The Compression Close capsule serves the following purposes. As a response to reject a COMPRESSION_ASSIGN request and to close or clean up any existing compression mappings. Once a COMPRESSION_CLOSE is sent for a given Context ID, the sending party MAY reject any datagrams received for that Context ID until it is reallocated through a COMPRESSION_ASSIGN exchange.
+
+~~~ ascii-art
+Capsule {
+  Type COMPRESSION_CLOSE (0x1F1F1F20),
+  Length (i),
+  Context ID (i),
+}
+~~~
+{: #capsulecloseformat title="Compression Close Capsule Format"}
+
+
+
+# The connect-udp-bind Header Field {#hdr}
+
+The "connect-udp-bind" header field’s value is a boolean structured field
```

Comment:
This paragraph should explain what the purpose of the header is (to enable this extension)